### PR TITLE
FEATURE: Add Date.formatCldr helper

### PR DESCRIPTION
**Pull request by zaveryukha**
This change adds CLDR formatting capability to the Date Eel helper, to match
the functionality of the `format.date` Fluid ViewHelper.

**P.S.** This PR has taken over from https://github.com/neos/flow-development-collection/pull/920
All the history can be found the. The original code was written by @litovchenko1, I
(@zaveryukha) have added the unit tests.

### CODE REVIEW

**Comment by gerhard-boden** [comment]
@zaveryukha Thanks a lot for taking over! Would you mind fixing the nitpicks from #920? 
This would go into the upcoming 4.2 release.

**Comment by zaveryukha** [comment]
@gerhard-boden I have read all the comments and looks like I have fixed all of them.

File: Neos.Eel/Classes/Helper/DateHelper.php [hunk 1]
```diff
@@ -50,13 +63,49 @@ public function format($date, $format)
...
+     * Format a date to a string with a given cldr format
+     *
+     * @param integer|string|\DateTime $date
+     * @param string $cldrFormat Format string in CLDR format (see http://cldr.unicode.org/translation/date-time)
+     * @param null|string $locale String locale - example (de|en|ru_RU)
+     * @return string
+     */
+    public function formatCldr($date, $cldrFormat, $locale = null)
+    {
+        if ($date instanceof \DateTimeInterface) {
```
Comment (dimaip) [inline Neos.Eel/Classes/Helper/DateHelper.php]: This condition is not needed

Comment (zaveryukha) [inline Neos.Eel/Classes/Helper/DateHelper.php]: @dimaip done

File: Neos.Eel/Tests/Unit/Helper/DateHelperTest.php [hunk 2]
```diff
@@ -52,7 +53,7 @@ public function formatExamples()
         return [
             'DateTime object' => [$dateTime, 'Y-m-d H:i:s', '2013-07-03 12:34:56'],
             'timestamp as integer' => [1372856513, 'Y-m-d', '2013-07-03'],
-            'timestamp as string' => ['1372856513', 'Y-m-d', '2013-07-03'],
+            'timestamp as string' => ['@1372856513', 'Y-m-d', '2013-07-03'],
```
Comment (albe) [inline Neos.Eel/Tests/Unit/Helper/DateHelperTest.php]: The change in behaviour here would make this change breaking. Can we make this b/c by checking a `string` argument if it `is_numeric`?

Comment (zaveryukha) [inline Neos.Eel/Tests/Unit/Helper/DateHelperTest.php]: @albe we discuss this with @kitsunet 
The problem what the same string '20160101' means completely different things if it date or timestamp. 
Timestamp as a string '20160101' is equivalent to: 22/08/1970 08:08:41 (format 'd/m/Y H:m:s')
Date as a string '20160101' is equivalent to: 01/01/2016 00:01:00 (format 'd/m/Y H:m:s')
IMO we must clearly separate these things.

Comment (kitsunet) [inline Neos.Eel/Tests/Unit/Helper/DateHelperTest.php]: I am somewhat fine with this breakiness TBH. But Up to other reviewers.

Comment (albe) [inline Neos.Eel/Tests/Unit/Helper/DateHelperTest.php]: Well, yeah, I'm basically fine with the change of behaviour per se, just not sure if we should add it in a minor release.

**Comment by kdambekalns** [comment]
Adding some of the things from the description I tweaked to fit our needs:

---

**How to verify it**

Just use `${Date.formatCldr(someDateObject, 'dd MMMM yyyy HH:mm:ss', 'de')}` to test. Also test different combinations with missing arguments.

**Checklist**

- [X] Code follows the PSR-2 coding style
- [X] Tests have been created, run and adjusted as needed
- [X] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

File: Neos.Eel/Tests/Unit/Helper/DateHelperTest.php (hunk 2 above)
Comment (kdambekalns) [inline Neos.Eel/Tests/Unit/Helper/DateHelperTest.php]: The `$time` argument to `DateTime::__construct()` **is a string** and to use a unix timestamp one has to use `@1234567`. `date()` on the other hand only accepts integers. 

So what we have here is a case of _our own syntactic sugar broke a defined behaviour for the sake of making things "easier"_…

IMHO whenever I have a timestamp it is a proper integer, so this change would not affect me. But I like the `is_numeric` suhggestion…

Comment (kdambekalns) [inline Neos.Eel/Tests/Unit/Helper/DateHelperTest.php]: Scrap that. Indeed a string like `20160101` is parsed as a date and would still be numeric, leading to very strange behaviour. Talk about magic…

So my vote: drop string handling. If people want to parse a string, why not let them use `DateTime` themselves? Using a proper object to pass dates around feels better anyway.

**Comment by gerhard-boden** [comment]
@zaveryukha I agree with Karten's request to drop the support for string to date altogether, since the behaviour is not really predictable.

The CLDR in Eel part would still be a nice feature for 4.2, would you mind removing the code portions?

**Comment by albe** [comment]
Haha, my comment on keeping it b/c somehow backfired :D I guess it makes sense to completely avoid the fuzzy behaviour of the DateTime constructor with strings.

I'd still prefer to introduce the breaking changes for next major only, but if we do now, let's mark this PR as breaking (!!!) and exactly document in the PR description what changed.

**Comment by dimaip** [comment]
Since it's extremely rare that we would need to render dates from strings and not from DateTime objects coming from DB/CR, I don't really mind how this helper will work, as long as the two helpers are consistent with each other.
@zaveryukha what do you think? Will you be able to finish this helper in the mean time or should someone take over?

**Comment by zaveryukha** [comment]
@dimaip I'll update it today. No problem.

**Comment by kdambekalns** [comment]
Well, for me "drop string handling" was more about the "new string handling", we can keep the cast in place so it behaves like before. Then drop it completely for the next major.

**Comment by zaveryukha** [comment]
I have removed string handling from the format helper.

**Comment by gerhard-boden** [comment]
@zaveryukha Thank you!
@kdambekalns Are you ready to remove your block on this one?

**Comment by gerhard-boden** [comment]
Thanks to everyone for the blazing fast reactions and special thanks to @zaveryukha and @litovchenko1 for their code contributions.  This will make it into 4.2

**Comment by dimaip** [comment]
Will just stop by here to congratulate the two new contributors!
Now Fusion got a little stronger :)

**Comment by kitsunet** [comment]
Indee, great work!

**Pull request merged**